### PR TITLE
Tooltip

### LINKED ISSUE #92: Tooltip widget
# Specification
This widget provides tooltip behaviour.
 
## Features
- You specify the content for a tooltip. 
- Tooltip content can contain arbitrary vDOM nodes and widgets, e.g. 
screen shot 2017-03-30 at 2 30 03 pm
- The tooltip can be positioned to open in different directions, _relative_ to the element that causes the tooltip to be shown, e.g. `{position: TooltipPosition.TopCenter}` would open a tooltip above the element.
screen shot 2017-03-30 at 2 32 50 pm

- You can specify the event that will launch a tooltip. Usually this will be the `hover` of an element, but it could also be a click event etc.

**Value-add of the widget (e.g. why use rather than just use VDom directly)**
Ability to provide advanced styling and animation.

**List of callback funcs that can be passed in props (if any)**
`onHide`
`onShow`

**Mouse/keyboard interactions (if any)**
Depends on the content of the tooltip. If the tooltip contained widgets, then would you need to be able to tab into this content - or is this _horrendous_ UX 🤔 This affects the acceptance criteria.

**Mandatory/valid/empty/wait states (if any)**
None

**Is the widget controlled/uncontrolled**
Controlled

**List of any icons needed**

**Design input required (inc responsive design)**
Look and feel of the tooltip, pointing arrow.

**Any other considerations**:

## Acceptance criteria
When I perform an action that should make a tooltip appear, the tooltip appears, without obscuring the element that I was interacting with.
When I perform an action that should make the tooltip disappear, then when the tooltip is no longer shown, the focus should be on the element that had focus before the tooltip was shown.

### PR DESCRIPTION
**Pull request by bitpshr**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

Tooltip

This PR introduces a new Tooltip component that can serve as a basis for other tooltip-related functionality like form component validation. The properties interface for Tooltip is intentionally minimal. Initially, children were wrapped in a single parent node, and specific events could be used to show or hide the tooltip (`click`, `mouseenter`, etc.) This seemed overly restrictive, especially when used in conjunction with other Dojo 2 widgets like a TextInput. Without mandating an API that all Tooltip targets / children must implement, and to support regular DOM, Dojo 2 components, and MyAmazingFramework widgets, the ability to show the tooltip must be fully inverted. Luckily for us, that keeps the code fairly clean.

Resolves #92

### CODE REVIEW

File: src/accordionpane/createAccordionPaneElement.ts [hunk 1]
```diff
@@ -8,8 +8,20 @@ export default function createAccordionPaneElement(): CustomElementDescriptor {
 	return {
 		tagName: 'dojo-accordionpane',
 		widgetConstructor: AccordionPane,
-		attributes: [],
-		properties: [],
-		events: []
+		properties: [
```
Comment (bitpshr) [inline src/accordionpane/createAccordionPaneElement.ts]: This was an omission on my part with the AccordionPane PR.

Comment (kitsonk) [reply src/accordionpane/createAccordionPaneElement.ts]: Why is it part of a tooltip PR?

File: src/slidepane/SlidePane.ts [hunk 2]
```diff
@@ -203,6 +203,8 @@ export default class SlidePane extends SlidePaneBase<SlidePaneProperties> {
 			align = Align.left,
 			open = false
 		} = this.properties;
+		// https://github.com/Microsoft/TypeScript/issues/13042
+		// const alignCss: {[Key in Align]: string} = css;
```
Comment (kitsonk) [inline src/slidepane/SlidePane.ts]: Even if there is an issue, an uncomment line of code is ugly... if there is further action here when the issue is resolved, we should have an issue created in widgets, referencing this issue, and detailing what we need to do so we don't forget about it.

Comment (kitsonk) [reply src/slidepane/SlidePane.ts]: Also again, seems wholly unrelated to this PR.

File: src/accordionpane/createAccordionPaneElement.ts (hunk 1 above)
Comment (bitpshr) [reply src/accordionpane/createAccordionPaneElement.ts]: Removed. Opened #364 to track adding this and the AccordionPane README.

File: src/slidepane/SlidePane.ts (hunk 2 above)
Comment (bitpshr) [reply src/slidepane/SlidePane.ts]: Agreed. I removed this comment altogether from SlidePane. I also created #365 to track the updating of component code once TS 2.7 is released and fixes the respective issue.

File: src/tooltip/README.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+Dojo 2's `Tooltip` component can be used to show or hide content around a child component based on specific events.
+
+## Features
+
+- Agnostic to an eventing implementation
+- Supports top, right, bottom, and left orientations
+- Optionally persistent
+
+![Image of basic dialog](http://placekitten.com/450/300)
```
Comment (tomdye) [inline src/tooltip/README.md]: not a dialog, remove placekitten

File: src/tooltip/README.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+*Basic Example*
+```typescript
+import TextInput from '@dojo/widgets/textinput/TextInput';
+import Tooltip from '@dojo/widgets/tooltip/Tooltip';
+import { w } from '@dojo/widget-core/d';
+
+w(Tooltip, {
+	key: 'foo',
+	content: 'This is a tooltip!',
+	showing: this._tooltipShowing
```
Comment (tomdye) [inline src/tooltip/README.md]: do we use `showing` or `show` / `open` / `visible` elsewhere? not a fan of the `showing` naming here

Comment (tomdye) [reply src/tooltip/README.md]: looks like `Dialog` and `SlidePane` use `open` ?

Comment (bitpshr) [reply src/tooltip/README.md]: A tooltip doesn't open or close, but I'll go with `open` for consistency.

**Pull request merged**